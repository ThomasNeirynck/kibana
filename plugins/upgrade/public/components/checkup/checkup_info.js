import PropTypes from 'prop-types';
import React from 'react';

import { Issues } from './issues';


const INFO_ISSUES = [
  {
    level: 'none',
    message: 'Everything is OK.',
  },
  {
    level: 'info',
    message: 'Info that something has changed. No action required to upgrade.',
  },
  {
    level: 'warning',
    message: (
      'Action recommended but not required to upgrade. You are using '
      + 'deprecated functionality which will not be available in version 6.0.'
    ),
  },
  {
    level: 'critical',
    message: 'Action required. You must fix this problem to upgrade.',
  },
];

export function CheckupInfo({ className }) {
  return (
    <div className={className}>
      <p className="kuiText kuiVerticalRhythm">
        This tool runs a series of checks against your Elasticsearch cluster, nodes, and indices
        to determine whether you can upgrade directly to Elasticsearch version 6, or whether you
        need to make changes to your data before doing so.
      </p>
      <p className="kuiText kuiVerticalRhythm">
        You will also see deprecated cluster settings and node settings, deprecated plugins, and
        indices with deprecated mappings currently in use.
      </p>
      <p className="kuiText kuiVerticalRhythm">
        Each issue found has a degree of severity:
      </p>
      <Issues className="kuiVerticalRhythm" issues={INFO_ISSUES} />
    </div>
  );
}

CheckupInfo.propTypes = {
  className: PropTypes.string,
};

CheckupInfo.defaultProps = {
  className: null,
};
