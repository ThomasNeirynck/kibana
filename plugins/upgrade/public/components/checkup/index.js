export { CheckupView } from './checkup_view';
