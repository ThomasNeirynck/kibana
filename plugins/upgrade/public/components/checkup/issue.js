import PropTypes from 'prop-types';
import React from 'react';

import { IssueSymbol } from './issue_symbol';


export function Issue({ details, level, message, url }) {
  return (
    <li className='kuiMenuItem'>
      <div className='kuiEvent'>
        <IssueSymbol level={ level } />

        <div className='kuiEventBody'>
          <div className='kuiEventBody__message'>
            { message }{ details ? `: ${details }` : null}
          </div>
          <div className='kuiEventBody__metadata'>
            {
              url
              ? <a className='kuiLink' href={ url }>View Documentation</a>
              : null
            }
          </div>
        </div>
      </div>
    </li>
  );
}

Issue.propTypes = {
  details: PropTypes.string,
  level: PropTypes.string,
  message: PropTypes.string,
  url: PropTypes.string,
};

Issue.defaultProps = {
  details: null,
  level: null,
  message: null,
  url: null,
};
