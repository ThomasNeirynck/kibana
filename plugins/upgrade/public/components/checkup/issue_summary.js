import PropTypes from 'prop-types';
import React from 'react';

import { DEPRECATION_ISSUE_LEVELS } from '../../lib/constants';


const ISSUE_LEVEL_MESSAGES = {
  info: (count) => (
    <span className="kuiStatusText kuiStatusText--info">
      <span className="kuiStatusText__icon kuiIcon fa-info" />
      {
        count === 1
        ? <span>{ count } note</span>
        : <span>{ count } notes</span>
      }
    </span>
  ),
  warning: (count) => (
    <span className="kuiStatusText kuiStatusText--warning">
      <span className="kuiStatusText__icon kuiIcon fa-bolt" />
      {
        count === 1
        ? <span>{ count } warning</span>
        : <span>{ count } warnings</span>
      }
    </span>
  ),
  critical: (count) => (
    <span className="kuiStatusText kuiStatusText--error">
      <span className="kuiStatusText__icon kuiIcon fa-warning" />
      {
        count === 1
        ? <span>{ count } error</span>
        : <span>{ count } errors</span>
      }
    </span>
  ),
  none: () => (
    <span className="kuiStatusText kuiStatusText--success">
      <span className="kuiStatusText__icon kuiIcon fa-check" />
      No problems
    </span>
  ),
};

export function IssueSummary({ className, issueLevelCounts }) {
  const highestIssueLevel = DEPRECATION_ISSUE_LEVELS.reduce((highestLevel, issueLevel) => {
    if (highestLevel !== 'none') {
      return highestLevel;
    }

    if ((issueLevelCounts[issueLevel] || 0) > 0) {
      return issueLevel;
    } else {
      return highestLevel;
    }
  }, 'none');

  return (
    <span className={className}>
      { ISSUE_LEVEL_MESSAGES[highestIssueLevel](issueLevelCounts[highestIssueLevel]) }
    </span>
  );
}

IssueSummary.propTypes = {
  className: PropTypes.string,
  issueLevelCounts: PropTypes.object,
};

IssueSummary.defaultProps = {
  className: null,
  issueLevelCounts: {},
};
