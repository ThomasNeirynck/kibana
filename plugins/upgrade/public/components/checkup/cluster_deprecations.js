import _ from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';

import { StatusGroup } from '../status_group';
import { Issues } from './issues';
import { IssueSummary } from './issue_summary';


export function ClusterDeprecations({ className, deprecations }) {
  const issueLevelCounts = _.countBy(deprecations, 'level');

  return (
    <StatusGroup
      className={className}
      isInitiallyCollapsed
      status={<IssueSummary issueLevelCounts={issueLevelCounts} />}
      title="Cluster Settings"
    >
      { deprecations.length > 0
        ? <Issues issues={deprecations} />
        : <p className="kuiText kuiSubduedText">No cluster settings deprecations</p>
      }
    </StatusGroup>
  );
}

ClusterDeprecations.propTypes = {
  className: PropTypes.string,
  deprecations: PropTypes.array,
};

ClusterDeprecations.defaultProps = {
  className: null,
  deprecations: [],
};
