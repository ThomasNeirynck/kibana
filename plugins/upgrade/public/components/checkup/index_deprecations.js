import _ from 'lodash';
import React from 'react';

import { StatusGroup } from '../status_group';
import { Issues } from './issues';
import { IssueSummary } from './issue_summary';


export function IndexDeprecations({ className, deprecations }) {
  const indexNames = Object.keys(deprecations);
  const levelCountsByIndex = _.reduce(deprecations, (result, levelCounts, indexName) => ({
    ...result,
    [indexName]: _.countBy(levelCounts, 'level'),
  }), {});
  const totalLevelCounts = _.reduce(levelCountsByIndex, (totalCounts, levelCounts) => ({
    none: totalCounts.none + (levelCounts.none || 0),
    info: totalCounts.info + (levelCounts.info || 0),
    warning: totalCounts.warning + (levelCounts.warning || 0),
    critical: totalCounts.critical + (levelCounts.critical || 0),
  }), {
    none: 0,
    info: 0,
    warning: 0,
    critical: 0,
  });

  return (
    <StatusGroup
      className={ className }
      isInitiallyCollapsed
      status={ <IssueSummary issueLevelCounts={ totalLevelCounts } /> }
      title="Index Settings"
    >
      { indexNames.length > 0
        ? (
          indexNames.map((indexName) => (
            <div key={ indexName }>
              <div className="kuiBar">
                <div className="kuiBarSection">
                  <p className='kuiTextTitle'>{ indexName }</p>
                </div>
                <div className="kuiBarSection">
                  <IssueSummary issueLevelCounts={ levelCountsByIndex[indexName] } />
                </div>
              </div>
              <Issues issues={ deprecations[indexName] } />
            </div>
          ))
        )
        : <p className="kuiText kuiSubduedText">No index-level deprecations</p>
      }
    </StatusGroup>
  );
}

IndexDeprecations.propTypes = {
  className: React.PropTypes.string,
  deprecations: React.PropTypes.objectOf(React.PropTypes.array),
};

IndexDeprecations.defaultProps = {
  className: null,
  deprecations: {},
};
