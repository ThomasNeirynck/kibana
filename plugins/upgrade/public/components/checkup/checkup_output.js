import PropTypes from 'prop-types';
import React from 'react';

import { IndexDeprecations } from './index_deprecations';
import { ClusterDeprecations } from './cluster_deprecations';
import { NodeDeprecations } from './node_deprecations';


export function CheckupOutput({ className, output }) {
  return (
    <div className={className}>
      <IndexDeprecations
        className="kuiVerticalRhythm"
        deprecations={output.index_settings}
      />
      <NodeDeprecations
        className="kuiVerticalRhythm"
        deprecations={output.node_settings}
      />
      <ClusterDeprecations
        className="kuiVerticalRhythm"
        deprecations={output.cluster_settings}
      />
    </div>
  );
}

CheckupOutput.propTypes = {
  className: PropTypes.string,
  output: PropTypes.shape({
    cluster_settings: PropTypes.array,
    index_settings: PropTypes.object,
    node_settings: PropTypes.array,
  }),
};

CheckupOutput.defaultProps = {
  className: null,
  output: {},
};
