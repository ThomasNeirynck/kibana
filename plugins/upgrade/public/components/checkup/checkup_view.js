import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { RefreshButton } from '../refresh_button';
import { CheckupOutput } from './checkup_output';
import { ErrorPanel } from '../error_panel';
import { CheckupInfo } from './checkup_info';
import { InfoGroup } from '../info_group';
import { LOADING_STATUS } from '../../lib/constants';
import { getDeprecations } from '../../lib';
import { withViewState } from '../../lib/util/view_state';


export const CheckupView = withViewState({
  initialState: {
    isInfoCollapsed: false,
  },
  updaters: {
    toggleInfoCollapsed: (state) => () => ({
      isInfoCollapsed: !state.isInfoCollapsed,
    }),
  },
})(class CheckupView extends Component {
  static propTypes = {
    isInfoCollapsed: PropTypes.bool,
    toggleInfoCollapsed: PropTypes.func,
  }

  static defaultProps = {
    isInfoCollapsed: false,
    toggleInfoCollapsed: () => {},
  }

  state = {
    deprecations: undefined,
    lastError: undefined,
    loadingStatus: LOADING_STATUS.UNINITIALIZED,
  };

  runCheckup = async () => {
    this.setState({
      loadingStatus: LOADING_STATUS.LOADING,
    });

    try {
      const deprecations = await getDeprecations();

      this.setState({
        deprecations,
        loadingStatus: LOADING_STATUS.SUCCESS,
      });
    } catch (error) {
      if (error.statusCode === 403) {
        this.setState({
          loadingStatus: LOADING_STATUS.FORBIDDEN,
        });
      } else {
        this.setState({
          lastError: error,
          loadingStatus: LOADING_STATUS.FAILURE,
        });
      }
    }
  }

  componentDidMount() {
    this.runCheckup();
  }

  render() {
    const { deprecations, lastError, loadingStatus } = this.state;
    const { isInfoCollapsed, toggleInfoCollapsed } = this.props;

    return (
      <div className="kuiView">
        <div className="kuiViewContent kuiViewContent--constrainedWidth">
          <div className="kuiViewContentItem">
            {
              loadingStatus === LOADING_STATUS.FORBIDDEN
                ? (
                  <ErrorPanel
                    className="kuiVerticalRhythm"
                    title="You do not have permission to use the Upgrade Assistant."
                  >
                    <p className="kuiText">Please contact your administrator.</p>
                  </ErrorPanel>
                )
                : (
                  <div className="kuiVerticalRhythm">
                    <InfoGroup
                      className="kuiVerticalRhythm"
                      isCollapsed={isInfoCollapsed}
                      onChangeCollapsed={toggleInfoCollapsed}
                      title="Cluster Checkup"
                    >
                      <CheckupInfo className="kuiVerticalRhythm" />
                    </InfoGroup>

                    <RefreshButton
                      buttonLabel="Rerun Checkup"
                      className="kuiVerticalRhythm"
                      onClick={this.runCheckup}
                    />
                  </div>
                )
            }

            {
              loadingStatus === LOADING_STATUS.FAILURE
                ? (
                  <ErrorPanel
                    className="kuiVerticalRhythm"
                    title="Failed to run checkup"
                  >
                    <p className="kuiText">{ lastError.message }</p>
                  </ErrorPanel>
                )
                : null
            }

            {
              loadingStatus === LOADING_STATUS.SUCCESS
                ? (
                  <CheckupOutput className="kuiVerticalRhythm" output={deprecations} />
                )
                : null
            }
          </div>
        </div>
      </div>
    );
  }
});
