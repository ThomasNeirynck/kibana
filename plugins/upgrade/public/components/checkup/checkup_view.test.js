import _ from 'lodash';
import { shallow } from 'enzyme';
import React from 'react';

import { CheckupView } from './checkup_view';
import { getDeprecations } from '../../lib/checkup';

jest.mock('ui/chrome', () => {}, { virtual: true });
jest.mock('ui/notify/notifier', () => ({
  Notifier: function Notifier() {},
}), { virtual: true });
jest.mock('../../lib/checkup');


describe('CheckupView', () => {
  afterEach(() => {
    getDeprecations.mockReset();
  });

  test('renders an expanded InfoGroup by default', () => {
    const component = (
      <CheckupView
        setViewState={ _.noop }
        viewState={ {} }
      />
    );

    const wrapper = shallow(component).dive();

    expect(wrapper.find('InfoGroup').prop('isCollapsed')).toBe(false);
  });

  test('renders a collapsed InfoGroup when the viewState declares it', () => {
    const component = (
      <CheckupView
        setViewState={ _.noop }
        viewState={ {
          CheckupView: {
            isInfoCollapsed: true,
          },
        } }
      />
    );

    const wrapper = shallow(component).dive();

    expect(wrapper.find('InfoGroup').prop('isCollapsed')).toBe(true);
  });

  test('calls setViewState when the InfoPanel\'s collapse button is clicked' , () => {
    const setViewState = jest.fn();
    const component = (
      <CheckupView
        setViewState={ setViewState }
        viewState={ {} }
      />
    );

    const wrapper = shallow(component).dive();
    wrapper.find('InfoGroup').prop('onChangeCollapsed')();

    expect(setViewState).toHaveBeenCalledTimes(1);
  });

  test('renders the CheckupOutput after successfully loading the deprecations', async () => {
    const responsePromise = Promise.resolve({
      cluster_settings: [],
      index_settings: {},
      node_settings: [],
    });
    getDeprecations.mockReturnValue(responsePromise);

    const component = (
      <CheckupView
        setViewState={ _.noop }
        viewState={ {} }
      />
    );

    const wrapper = shallow(component).dive();
    wrapper.instance().componentDidMount();
    await responsePromise;

    expect(getDeprecations).toHaveBeenCalledWith();
    const outputProp = wrapper.find('CheckupOutput').prop('output');
    expect(outputProp).toHaveProperty('cluster_settings');
    expect(outputProp).toHaveProperty('index_settings');
    expect(outputProp).toHaveProperty('node_settings');
  });

  test('renders an ErrorPanel after unsuccessfully trying to load the deprecations', async () => {
    const responsePromise = Promise.reject({
      message: 'an error'
    });
    getDeprecations.mockReturnValue(responsePromise);

    const component = (
      <CheckupView
        setViewState={ _.noop }
        viewState={ {} }
      />
    );

    const wrapper = shallow(component).dive();
    wrapper.instance().componentDidMount();
    await expect(responsePromise).rejects.toBeDefined;

    expect(getDeprecations).toHaveBeenCalledWith();
    expect(wrapper.find('ErrorPanel').contains('an error')).toBe(true);
  });

  test('renders a RefreshButton that reloads the deprecations', async () => {
    const responsePromise = Promise.resolve({
      cluster_settings: [],
      index_settings: {},
      node_settings: [],
    });
    getDeprecations.mockReturnValue(responsePromise);

    const component = (
      <CheckupView
        setViewState={ _.noop }
        viewState={ {} }
      />
    );

    const wrapper = shallow(component).dive();

    expect(wrapper.find('CheckupOutput').exists()).toBe(false);

    expect(wrapper.find('RefreshButton').exists()).toBe(true);
    wrapper.find('RefreshButton').simulate('click');
    await responsePromise;

    expect(getDeprecations).toHaveBeenCalledWith();
    expect(wrapper.find('CheckupOutput').exists()).toBe(true);
  });
});
