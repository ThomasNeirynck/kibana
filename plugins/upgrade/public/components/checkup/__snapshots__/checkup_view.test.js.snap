// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckupView renders an ErrorPanel when the loadingStatus indicates forbidden access 1`] = `undefined`;
